### PR TITLE
Correctly generate attachment URLs for non-winning rev

### PR DESCRIPTION
**Pull request by rhyshort**
Correctly generate attachment URLs for non-winning rev
by manually creating a  URL using URL components.

BugzId: 47846

reviewer @indisoluble 
reviewer @tomblench

### CODE REVIEW

**Comment by tomblench** [comment]
Tests? Also what's the behaviour with basic auth?

File: Classes/common/CDTDocumentRevision.m [hunk 1]
```diff
@@ -79,11 +79,17 @@ + (CDTDocumentRevision *)createRevisionFromJson:(NSDictionary *)jsonDict
 
     // build the attachment objects
     for (NSString *key in [attachmentData allKeys]) {
-        NSURL *url = [NSURL URLWithString:key relativeToURL:documentURL];
+        
+        NSURLComponents *documentUrlComponents = [NSURLComponents componentsWithURL:documentURL resolvingAgainstBaseURL:NO];
+        NSString *documentPathString = documentUrlComponents.path;
+        NSString *attachmentPath = [NSString stringWithFormat:@"%@/%@", documentPathString,key];
+        documentUrlComponents.path = attachmentPath;
```
Comment (indisoluble) [inline Classes/common/CDTDocumentRevision.m]: Based on the description in BugzId 47846, I think the problem was that the URL did not include the `<REVID>`.
Where are you adding the `<REVID>`? I'm not able to see it.

**Comment by indisoluble** [comment]
I agree with @tomblench, at least one test for the new code would be appreciated.

File: Classes/common/CDTDocumentRevision.m (hunk 1 above)
Comment (rhyshort) [inline Classes/common/CDTDocumentRevision.m]: the query is already in the `NSURLComponents` object, we are just replacing the path

**Comment by rhyshort** [comment]
I've yet to figure out how to test it since the URL is an internal property and the suite the tests for this are part of are don't use a couch server

File: Classes/common/CDTDocumentRevision.m (hunk 1 above)
Comment (indisoluble) [inline Classes/common/CDTDocumentRevision.m]: Thanks.

**Comment by tomblench** [comment]
@rhyshort it should be easy enough to write a test in acceptance tests which runs against local couch, or am I missing something? Although you can't directly check the URL you can test that createRevisionFromJson picks up the right rev and therefore you know it generated the right URL.

**Comment by rhyshort** [comment]
@tomblench I can't just check the URL generated since its private on the CDTSavedHTTPAttachment class unless I do some digging with the objective c runtime

File: ReplicationAcceptance/ReplicationAcceptance/Attachments.m [hunk 2]
```diff
@@ -100,6 +100,118 @@ - (void)deleteAttachmentNamed:(NSString*)attachmentName
...
+    NSMutableDictionary *attachments = [NSMutableDictionary dictionary];
+        NSString *content = @"blahblah";
+        NSData *data = [content dataUsingEncoding:NSUTF8StringEncoding];
+        NSString *name = @"attachment";
+        CDTAttachment *attachment = [[CDTUnsavedDataAttachment alloc] initWithData:data
+                                                                              name:name
+                                                                              type:@"text/plain"];
+        [attachments setObject:attachment forKey:name];
+        
+        [originalAttachments setObject:data forKey:name];
```
Comment (indisoluble) [inline ReplicationAcceptance/ReplicationAcceptance/Attachments.m]: These lines are not correctly tabulated.

Comment (rhyshort) [inline ReplicationAcceptance/ReplicationAcceptance/Attachments.m]: done in 1748ba8

**Comment by indisoluble** [comment]
:+1:

File: ReplicationAcceptance/ReplicationAcceptance/Attachments.m [hunk 3]
```diff
@@ -100,6 +100,118 @@ - (void)deleteAttachmentNamed:(NSString*)attachmentName
...
+    XCTAssertTrue([self compareDatastore:self.datastore withDatabase:remoteDbURL],
+                  @"Local and remote database comparison failed");
+    
+    XCTAssertTrue([self compareAttachmentsForCurrentRevisions:self.datastore
+                                                 withDatabase:remoteDbURL],
+                  @"Local and remote database attachment comparison failed");
+    
+    NSDictionary *responseDict = response.body.JSONObject;
+    
+    CDTDocumentRevision *remoteRevision = [CDTDocumentRevision createRevisionFromJson:responseDict forDocument:docURL error:&error];
```
Comment (tomblench) [inline ReplicationAcceptance/ReplicationAcceptance/Attachments.m]: Assert on remoteRevision revid?

**Comment by tomblench** [comment]
+1

File: ReplicationAcceptance/ReplicationAcceptance/Attachments.m (hunk 3 above)
Comment (rhyshort) [inline ReplicationAcceptance/ReplicationAcceptance/Attachments.m]: I wouldn't say its needed since the attachments are deleted on the current winning rev.

**Pull request merged**